[kjobctl] Add describe slurm

IrvingMg opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature

#### What this PR does / why we need it:

Add `describe` slurm command to kjobctl.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

mbobrovskyi commented:
Can you add tests?

IrvingMg commented:
/test all

IrvingMg commented:
/cc @mbobrovskyi

mbobrovskyi commented on cmd/experimental/kjobctl/pkg/cmd/describe/describe_printer.go:
```diff
@@ -40,7 +41,7 @@ type ResourceDescriber interface {
 }
 
-// NewResourceDescriber Describer returns a Describer for displaying the specified RESTMapping type or an error.
+// Describer returns a Describer for displaying the specified RESTMapping type or an error.
```
```suggestion
// NewResourceDescriber returns a Describer for displaying the specified RESTMapping type or an error.
```

mbobrovskyi commented:
/lgtm

alculquicondor commented on cmd/experimental/kjobctl/pkg/cmd/describe/describe.go:
```diff
@@ -261,3 +284,57 @@ func (o *DescribeOptions) customizeResourceBuilder() *resource.Builder {
+		infos = append(infos, info)
+	} else {
+		cmList, err := o.Clientset.CoreV1().ConfigMaps(o.Namespace).List(ctx, metav1.ListOptions{
+			LabelSelector: constants.ProfileLabel,
```
isnt' this getting all the configmaps that have a profile label?

How is this specific to one job?

IrvingMg replied:
The `describe` command allows to see a specific job or all the jobs created for a specific mode. It's for that second case when we need to retrieve all the configmaps created for Slurm mode.

However, I was missing filtering the configmaps by Mode label as well. I´ve updated the label selector to include it.

alculquicondor replied:
Will the config maps be printed alongside their respective job?
It looks like jobs will be printed first, then all the configmaps

IrvingMg replied:
Yes, currently jobs are printed first, and then the configmaps

alculquicondor replied:
Can we interleave them so that jobs are always next to their configmaps?
Otherwise the output is not very usable.

alculquicondor commented on cmd/experimental/kjobctl/pkg/cmd/describe/describe.go:
```diff
@@ -198,6 +199,10 @@ func (o *DescribeOptions) Run(ctx context.Context) error {
 	}
 
+	slices.SortFunc(infos, func(a, b *resource.Info) int {
+		return strings.Compare(a.Name, b.Name)
```
is this enough guarantee? Are the names always supposed to match?

mbobrovskyi replied:
Yes, we are creating `Job` and `ConfigMap` with the same name.

alculquicondor replied:
Add a comment to the function indicating why we are sorting and why it works.

mwielgus commented:
Can you paste a sample command output?

IrvingMg commented:
@mwielgus 

Sample 1
---
Input: 
```
describe slurm sample-profile-slurm-m2lgd
```

Output:
```
Name:           sample-profile-slurm-m2lgd
Namespace:      default
Labels:         kjobctl.x-k8s.io/mode=Slurm
                kjobctl.x-k8s.io/profile=sample-profile
Parallelism:    2
Completions:    3
Start Time:     Tue, 10 Sep 2024 17:12:31 +0200
Pods Statuses:  2 Active (0 Ready) / 0 Succeeded / 0 Failed
Pod Template:
  Containers:
   sample-container:
    Port:       <none>
    Host Port:  <none>
    Command:
      bash
      /slurm/entrypoint.sh
    Environment:
      ENTRYPOINT_PATH:      /home/slurm/samples
      USER_ID:              Irving_Mondragon
      TASK_NAME:            default_sample-profile
      TASK_ID:              Irving_Mondragon_2024-09-10T17:12:31+02:00_default_sample-profile
      PROFILE:              default_sample-profile
      TIMESTAMP:            2024-09-10T17:12:31+02:00
      JOB_CONTAINER_INDEX:  0
    Mounts:
      /home/slurm/samples from slurm-code-sample (rw)
      /slurm from sample-profile-slurm-m2lgd (rw)
  Volumes:
   slurm-code-sample:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      slurm-code-sample
    Optional:  false
   sample-profile-slurm-m2lgd:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      sample-profile-slurm-m2lgd
    Optional:  false


Name:       sample-profile-slurm-m2lgd
Namespace:  default
Labels:     kjobctl.x-k8s.io/mode=Slurm
            kjobctl.x-k8s.io/profile=sample-profile

Data
====
entrypoint.sh:
----
#!/usr/bin/bash

set -o errexit
set -o nounset
set -o pipefail

# External variables
# JOB_COMPLETION_INDEX  - completion index of the job.
# JOB_CONTAINER_INDEX   - container index in the container template.

# ["COMPLETION_INDEX"]="CONTAINER_INDEX_1,CONTAINER_INDEX_2"
declare -A array_indexes=(["0"]="1" ["1"]="2" ["2"]="3")   # Requires bash v4+

container_indexes=${array_indexes[${JOB_COMPLETION_INDEX}]}
container_indexes=(${container_indexes//,/ })

if [[ ! -v container_indexes[${JOB_CONTAINER_INDEX}] ]];
then
  exit 0
fi

SBATCH_ARRAY_INX=1-3%2
SBATCH_GPUS_PER_TASK=
SBATCH_MEM_PER_CPU=
SBATCH_MEM_PER_GPU=
SBATCH_OUTPUT=
SBATCH_ERROR=
SBATCH_INPUT=
SBATCH_JOB_NAME=
SBATCH_PARTITION=

export SLURM_ARRAY_JOB_ID=1                  # Job array’s master job ID number.
export SLURM_ARRAY_TASK_COUNT=3              # Total number of tasks in a job array.
export SLURM_ARRAY_TASK_MAX=3                # Job array’s maximum ID (index) number.
export SLURM_ARRAY_TASK_MIN=1                # Job array’s minimum ID (index) number.
export SLURM_TASKS_PER_NODE=1                # Number of tasks to be initiated on each node.
export SLURM_CPUS_PER_TASK=                  # Number of CPUs per task.
export SLURM_CPUS_ON_NODE=                   # Number of CPUs on the allocated node (actually pod).
export SLURM_JOB_CPUS_PER_NODE=              # Count of processors available to the job on this node.
export SLURM_CPUS_PER_GPU=                   # Number of CPUs requested per allocated GPU.
export SLURM_MEM_PER_CPU=                  # Memory per CPU. Same as --mem-per-cpu .
export SLURM_MEM_PER_GPU=                  # Memory per GPU.
export SLURM_MEM_PER_NODE=                 # Memory per node. Same as --mem.
export SLURM_GPUS=0                        # Number of GPUs requested (in total).
export SLURM_NTASKS=1                      # Same as -n, –ntasks. The number of tasks.
export SLURM_NTASKS_PER_NODE=1               # Number of tasks requested per node.
export SLURM_NPROCS=$SLURM_NTASKS          # Same as -n, --ntasks. See $SLURM_NTASKS.
export SLURM_NNODES=2                        # Total number of nodes (actually pods) in the job’s resource allocation.
export SLURM_SUBMIT_DIR=/slurm               # The path of the job submission directory.
export SLURM_SUBMIT_HOST=$HOSTNAME         # The hostname of the node used for job submission.

export SLURM_JOB_ID=$(( JOB_COMPLETION_INDEX * SLURM_TASKS_PER_NODE + JOB_CONTAINER_INDEX + SLURM_ARRAY_JOB_ID ))   # The Job ID.
export SLURM_JOBID=$SLURM_JOB_ID                                                                                    # Deprecated. Same as $SLURM_JOB_ID
export SLURM_ARRAY_TASK_ID=${container_indexes[${JOB_CONTAINER_INDEX}]}                        # Task ID.

unmask_filename () {
  replaced="$1"

  if [[ "$replaced" == "\\"* ]]; then
      replaced="${replaced//\\/}"
      echo "${replaced}"
      return 0
  fi

  replaced=$(echo "$replaced" | sed -E "s/(%)(%A)/\1\n\2/g;:a s/(^|[^\n])%A/\1$SLURM_ARRAY_JOB_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%a)/\1\n\2/g;:a s/(^|[^\n])%a/\1$SLURM_ARRAY_TASK_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%j)/\1\n\2/g;:a s/(^|[^\n])%j/\1$SLURM_JOB_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%N)/\1\n\2/g;:a s/(^|[^\n])%N/\1$HOSTNAME/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%n)/\1\n\2/g;:a s/(^|[^\n])%n/\1$JOB_COMPLETION_INDEX/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%t)/\1\n\2/g;:a s/(^|[^\n])%t/\1$SLURM_ARRAY_TASK_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%u)/\1\n\2/g;:a s/(^|[^\n])%u/\1$USER_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%x)/\1\n\2/g;:a s/(^|[^\n])%x/\1$SBATCH_JOB_NAME/;ta;s/\n//g")

  replaced="${replaced//%%/%}"

  echo "$replaced"
}

input_file=$(unmask_filename "$SBATCH_INPUT")
output_file=$(unmask_filename "$SBATCH_OUTPUT")
error_path=$(unmask_filename "$SBATCH_ERROR")

/slurm/script


script:
----
#!/bin/bash

#SBATCH --array=1-3%2

echo "now processing task id:: ${SLURM_ARRAY_TASK_ID}"
python /home/slurm/samples/sample_code.py

exit 0



BinaryData
====
```

Sample 2
---
Input:
```
describe slurm
```

Output:
```
Name:           sample-profile-slurm-m2lgd
Namespace:      default
Labels:         kjobctl.x-k8s.io/mode=Slurm
                kjobctl.x-k8s.io/profile=sample-profile
Parallelism:    2
Completions:    3
Start Time:     Tue, 10 Sep 2024 17:12:31 +0200
Pods Statuses:  2 Active (0 Ready) / 0 Succeeded / 0 Failed
Pod Template:
  Containers:
   sample-container:
    Port:       <none>
    Host Port:  <none>
    Command:
      bash
      /slurm/entrypoint.sh
    Environment:
      ENTRYPOINT_PATH:      /home/slurm/samples
      USER_ID:              Irving_Mondragon
      TASK_NAME:            default_sample-profile
      TASK_ID:              Irving_Mondragon_2024-09-10T17:12:31+02:00_default_sample-profile
      PROFILE:              default_sample-profile
      TIMESTAMP:            2024-09-10T17:12:31+02:00
      JOB_CONTAINER_INDEX:  0
    Mounts:
      /home/slurm/samples from slurm-code-sample (rw)
      /slurm from sample-profile-slurm-m2lgd (rw)
  Volumes:
   slurm-code-sample:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      slurm-code-sample
    Optional:  false
   sample-profile-slurm-m2lgd:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      sample-profile-slurm-m2lgd
    Optional:  false


Name:       sample-profile-slurm-m2lgd
Namespace:  default
Labels:     kjobctl.x-k8s.io/mode=Slurm
            kjobctl.x-k8s.io/profile=sample-profile

Data
====
entrypoint.sh:
----
#!/usr/bin/bash

set -o errexit
set -o nounset
set -o pipefail

# External variables
# JOB_COMPLETION_INDEX  - completion index of the job.
# JOB_CONTAINER_INDEX   - container index in the container template.

# ["COMPLETION_INDEX"]="CONTAINER_INDEX_1,CONTAINER_INDEX_2"
declare -A array_indexes=(["0"]="1" ["1"]="2" ["2"]="3")   # Requires bash v4+

container_indexes=${array_indexes[${JOB_COMPLETION_INDEX}]}
container_indexes=(${container_indexes//,/ })

if [[ ! -v container_indexes[${JOB_CONTAINER_INDEX}] ]];
then
  exit 0
fi

SBATCH_ARRAY_INX=1-3%2
SBATCH_GPUS_PER_TASK=
SBATCH_MEM_PER_CPU=
SBATCH_MEM_PER_GPU=
SBATCH_OUTPUT=
SBATCH_ERROR=
SBATCH_INPUT=
SBATCH_JOB_NAME=
SBATCH_PARTITION=

export SLURM_ARRAY_JOB_ID=1                  # Job array’s master job ID number.
export SLURM_ARRAY_TASK_COUNT=3              # Total number of tasks in a job array.
export SLURM_ARRAY_TASK_MAX=3                # Job array’s maximum ID (index) number.
export SLURM_ARRAY_TASK_MIN=1                # Job array’s minimum ID (index) number.
export SLURM_TASKS_PER_NODE=1                # Number of tasks to be initiated on each node.
export SLURM_CPUS_PER_TASK=                  # Number of CPUs per task.
export SLURM_CPUS_ON_NODE=                   # Number of CPUs on the allocated node (actually pod).
export SLURM_JOB_CPUS_PER_NODE=              # Count of processors available to the job on this node.
export SLURM_CPUS_PER_GPU=                   # Number of CPUs requested per allocated GPU.
export SLURM_MEM_PER_CPU=                  # Memory per CPU. Same as --mem-per-cpu .
export SLURM_MEM_PER_GPU=                  # Memory per GPU.
export SLURM_MEM_PER_NODE=                 # Memory per node. Same as --mem.
export SLURM_GPUS=0                        # Number of GPUs requested (in total).
export SLURM_NTASKS=1                      # Same as -n, –ntasks. The number of tasks.
export SLURM_NTASKS_PER_NODE=1               # Number of tasks requested per node.
export SLURM_NPROCS=$SLURM_NTASKS          # Same as -n, --ntasks. See $SLURM_NTASKS.
export SLURM_NNODES=2                        # Total number of nodes (actually pods) in the job’s resource allocation.
export SLURM_SUBMIT_DIR=/slurm               # The path of the job submission directory.
export SLURM_SUBMIT_HOST=$HOSTNAME         # The hostname of the node used for job submission.

export SLURM_JOB_ID=$(( JOB_COMPLETION_INDEX * SLURM_TASKS_PER_NODE + JOB_CONTAINER_INDEX + SLURM_ARRAY_JOB_ID ))   # The Job ID.
export SLURM_JOBID=$SLURM_JOB_ID                                                                                    # Deprecated. Same as $SLURM_JOB_ID
export SLURM_ARRAY_TASK_ID=${container_indexes[${JOB_CONTAINER_INDEX}]}                        # Task ID.

unmask_filename () {
  replaced="$1"

  if [[ "$replaced" == "\\"* ]]; then
      replaced="${replaced//\\/}"
      echo "${replaced}"
      return 0
  fi

  replaced=$(echo "$replaced" | sed -E "s/(%)(%A)/\1\n\2/g;:a s/(^|[^\n])%A/\1$SLURM_ARRAY_JOB_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%a)/\1\n\2/g;:a s/(^|[^\n])%a/\1$SLURM_ARRAY_TASK_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%j)/\1\n\2/g;:a s/(^|[^\n])%j/\1$SLURM_JOB_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%N)/\1\n\2/g;:a s/(^|[^\n])%N/\1$HOSTNAME/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%n)/\1\n\2/g;:a s/(^|[^\n])%n/\1$JOB_COMPLETION_INDEX/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%t)/\1\n\2/g;:a s/(^|[^\n])%t/\1$SLURM_ARRAY_TASK_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%u)/\1\n\2/g;:a s/(^|[^\n])%u/\1$USER_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%x)/\1\n\2/g;:a s/(^|[^\n])%x/\1$SBATCH_JOB_NAME/;ta;s/\n//g")

  replaced="${replaced//%%/%}"

  echo "$replaced"
}

input_file=$(unmask_filename "$SBATCH_INPUT")
output_file=$(unmask_filename "$SBATCH_OUTPUT")
error_path=$(unmask_filename "$SBATCH_ERROR")

/slurm/script


script:
----
#!/bin/bash

#SBATCH --array=1-3%2

echo "now processing task id:: ${SLURM_ARRAY_TASK_ID}"
python /home/slurm/samples/sample_code.py

exit 0



BinaryData
====



Name:           sample-profile-slurm-qzcv4
Namespace:      default
Labels:         kjobctl.x-k8s.io/mode=Slurm
                kjobctl.x-k8s.io/profile=sample-profile
Parallelism:    2
Completions:    3
Start Time:     Tue, 10 Sep 2024 17:11:41 +0200
Pods Statuses:  2 Active (0 Ready) / 0 Succeeded / 0 Failed
Pod Template:
  Containers:
   sample-container:
    Port:       <none>
    Host Port:  <none>
    Command:
      bash
      /slurm/entrypoint.sh
    Environment:
      ENTRYPOINT_PATH:      /home/slurm/samples
      USER_ID:              Irving_Mondragon
      TASK_NAME:            default_sample-profile
      TASK_ID:              Irving_Mondragon_2024-09-10T17:11:41+02:00_default_sample-profile
      PROFILE:              default_sample-profile
      TIMESTAMP:            2024-09-10T17:11:41+02:00
      JOB_CONTAINER_INDEX:  0
    Mounts:
      /home/slurm/samples from slurm-code-sample (rw)
      /slurm from sample-profile-slurm-qzcv4 (rw)
  Volumes:
   slurm-code-sample:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      slurm-code-sample
    Optional:  false
   sample-profile-slurm-qzcv4:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      sample-profile-slurm-qzcv4
    Optional:  false


Name:       sample-profile-slurm-qzcv4
Namespace:  default
Labels:     kjobctl.x-k8s.io/mode=Slurm
            kjobctl.x-k8s.io/profile=sample-profile

Data
====
entrypoint.sh:
----
#!/usr/bin/bash

set -o errexit
set -o nounset
set -o pipefail

# External variables
# JOB_COMPLETION_INDEX  - completion index of the job.
# JOB_CONTAINER_INDEX   - container index in the container template.

# ["COMPLETION_INDEX"]="CONTAINER_INDEX_1,CONTAINER_INDEX_2"
declare -A array_indexes=(["0"]="1" ["1"]="2" ["2"]="3")   # Requires bash v4+

container_indexes=${array_indexes[${JOB_COMPLETION_INDEX}]}
container_indexes=(${container_indexes//,/ })

if [[ ! -v container_indexes[${JOB_CONTAINER_INDEX}] ]];
then
  exit 0
fi

SBATCH_ARRAY_INX=1-3%2
SBATCH_GPUS_PER_TASK=
SBATCH_MEM_PER_CPU=
SBATCH_MEM_PER_GPU=
SBATCH_OUTPUT=
SBATCH_ERROR=
SBATCH_INPUT=
SBATCH_JOB_NAME=
SBATCH_PARTITION=

export SLURM_ARRAY_JOB_ID=1                  # Job array’s master job ID number.
export SLURM_ARRAY_TASK_COUNT=3              # Total number of tasks in a job array.
export SLURM_ARRAY_TASK_MAX=3                # Job array’s maximum ID (index) number.
export SLURM_ARRAY_TASK_MIN=1                # Job array’s minimum ID (index) number.
export SLURM_TASKS_PER_NODE=1                # Number of tasks to be initiated on each node.
export SLURM_CPUS_PER_TASK=                  # Number of CPUs per task.
export SLURM_CPUS_ON_NODE=                   # Number of CPUs on the allocated node (actually pod).
export SLURM_JOB_CPUS_PER_NODE=              # Count of processors available to the job on this node.
export SLURM_CPUS_PER_GPU=                   # Number of CPUs requested per allocated GPU.
export SLURM_MEM_PER_CPU=                  # Memory per CPU. Same as --mem-per-cpu .
export SLURM_MEM_PER_GPU=                  # Memory per GPU.
export SLURM_MEM_PER_NODE=                 # Memory per node. Same as --mem.
export SLURM_GPUS=0                        # Number of GPUs requested (in total).
export SLURM_NTASKS=1                      # Same as -n, –ntasks. The number of tasks.
export SLURM_NTASKS_PER_NODE=1               # Number of tasks requested per node.
export SLURM_NPROCS=$SLURM_NTASKS          # Same as -n, --ntasks. See $SLURM_NTASKS.
export SLURM_NNODES=2                        # Total number of nodes (actually pods) in the job’s resource allocation.
export SLURM_SUBMIT_DIR=/slurm               # The path of the job submission directory.
export SLURM_SUBMIT_HOST=$HOSTNAME         # The hostname of the node used for job submission.

export SLURM_JOB_ID=$(( JOB_COMPLETION_INDEX * SLURM_TASKS_PER_NODE + JOB_CONTAINER_INDEX + SLURM_ARRAY_JOB_ID ))   # The Job ID.
export SLURM_JOBID=$SLURM_JOB_ID                                                                                    # Deprecated. Same as $SLURM_JOB_ID
export SLURM_ARRAY_TASK_ID=${container_indexes[${JOB_CONTAINER_INDEX}]}                        # Task ID.

unmask_filename () {
  replaced="$1"

  if [[ "$replaced" == "\\"* ]]; then
      replaced="${replaced//\\/}"
      echo "${replaced}"
      return 0
  fi

  replaced=$(echo "$replaced" | sed -E "s/(%)(%A)/\1\n\2/g;:a s/(^|[^\n])%A/\1$SLURM_ARRAY_JOB_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%a)/\1\n\2/g;:a s/(^|[^\n])%a/\1$SLURM_ARRAY_TASK_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%j)/\1\n\2/g;:a s/(^|[^\n])%j/\1$SLURM_JOB_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%N)/\1\n\2/g;:a s/(^|[^\n])%N/\1$HOSTNAME/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%n)/\1\n\2/g;:a s/(^|[^\n])%n/\1$JOB_COMPLETION_INDEX/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%t)/\1\n\2/g;:a s/(^|[^\n])%t/\1$SLURM_ARRAY_TASK_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%u)/\1\n\2/g;:a s/(^|[^\n])%u/\1$USER_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%x)/\1\n\2/g;:a s/(^|[^\n])%x/\1$SBATCH_JOB_NAME/;ta;s/\n//g")

  replaced="${replaced//%%/%}"

  echo "$replaced"
}

input_file=$(unmask_filename "$SBATCH_INPUT")
output_file=$(unmask_filename "$SBATCH_OUTPUT")
error_path=$(unmask_filename "$SBATCH_ERROR")

/slurm/script


script:
----
#!/bin/bash

#SBATCH --array=1-3%2

echo "now processing task id:: ${SLURM_ARRAY_TASK_ID}"
python /home/slurm/samples/sample_code.py

exit 0



BinaryData
====



Name:           sample-profile-slurm-vscv7
Namespace:      default
Labels:         kjobctl.x-k8s.io/mode=Slurm
                kjobctl.x-k8s.io/profile=sample-profile
Parallelism:    2
Completions:    3
Start Time:     Tue, 10 Sep 2024 17:12:41 +0200
Pods Statuses:  2 Active (0 Ready) / 0 Succeeded / 0 Failed
Pod Template:
  Containers:
   sample-container:
    Port:       <none>
    Host Port:  <none>
    Command:
      bash
      /slurm/entrypoint.sh
    Environment:
      ENTRYPOINT_PATH:      /home/slurm/samples
      USER_ID:              Irving_Mondragon
      TASK_NAME:            default_sample-profile
      TASK_ID:              Irving_Mondragon_2024-09-10T17:12:41+02:00_default_sample-profile
      PROFILE:              default_sample-profile
      TIMESTAMP:            2024-09-10T17:12:41+02:00
      JOB_CONTAINER_INDEX:  0
    Mounts:
      /home/slurm/samples from slurm-code-sample (rw)
      /slurm from sample-profile-slurm-vscv7 (rw)
  Volumes:
   slurm-code-sample:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      slurm-code-sample
    Optional:  false
   sample-profile-slurm-vscv7:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      sample-profile-slurm-vscv7
    Optional:  false


Name:       sample-profile-slurm-vscv7
Namespace:  default
Labels:     kjobctl.x-k8s.io/mode=Slurm
            kjobctl.x-k8s.io/profile=sample-profile

Data
====
entrypoint.sh:
----
#!/usr/bin/bash

set -o errexit
set -o nounset
set -o pipefail

# External variables
# JOB_COMPLETION_INDEX  - completion index of the job.
# JOB_CONTAINER_INDEX   - container index in the container template.

# ["COMPLETION_INDEX"]="CONTAINER_INDEX_1,CONTAINER_INDEX_2"
declare -A array_indexes=(["0"]="1" ["1"]="2" ["2"]="3")   # Requires bash v4+

container_indexes=${array_indexes[${JOB_COMPLETION_INDEX}]}
container_indexes=(${container_indexes//,/ })

if [[ ! -v container_indexes[${JOB_CONTAINER_INDEX}] ]];
then
  exit 0
fi

SBATCH_ARRAY_INX=1-3%2
SBATCH_GPUS_PER_TASK=
SBATCH_MEM_PER_CPU=
SBATCH_MEM_PER_GPU=
SBATCH_OUTPUT=
SBATCH_ERROR=
SBATCH_INPUT=
SBATCH_JOB_NAME=
SBATCH_PARTITION=

export SLURM_ARRAY_JOB_ID=1                  # Job array’s master job ID number.
export SLURM_ARRAY_TASK_COUNT=3              # Total number of tasks in a job array.
export SLURM_ARRAY_TASK_MAX=3                # Job array’s maximum ID (index) number.
export SLURM_ARRAY_TASK_MIN=1                # Job array’s minimum ID (index) number.
export SLURM_TASKS_PER_NODE=1                # Number of tasks to be initiated on each node.
export SLURM_CPUS_PER_TASK=                  # Number of CPUs per task.
export SLURM_CPUS_ON_NODE=                   # Number of CPUs on the allocated node (actually pod).
export SLURM_JOB_CPUS_PER_NODE=              # Count of processors available to the job on this node.
export SLURM_CPUS_PER_GPU=                   # Number of CPUs requested per allocated GPU.
export SLURM_MEM_PER_CPU=                  # Memory per CPU. Same as --mem-per-cpu .
export SLURM_MEM_PER_GPU=                  # Memory per GPU.
export SLURM_MEM_PER_NODE=                 # Memory per node. Same as --mem.
export SLURM_GPUS=0                        # Number of GPUs requested (in total).
export SLURM_NTASKS=1                      # Same as -n, –ntasks. The number of tasks.
export SLURM_NTASKS_PER_NODE=1               # Number of tasks requested per node.
export SLURM_NPROCS=$SLURM_NTASKS          # Same as -n, --ntasks. See $SLURM_NTASKS.
export SLURM_NNODES=2                        # Total number of nodes (actually pods) in the job’s resource allocation.
export SLURM_SUBMIT_DIR=/slurm               # The path of the job submission directory.
export SLURM_SUBMIT_HOST=$HOSTNAME         # The hostname of the node used for job submission.

export SLURM_JOB_ID=$(( JOB_COMPLETION_INDEX * SLURM_TASKS_PER_NODE + JOB_CONTAINER_INDEX + SLURM_ARRAY_JOB_ID ))   # The Job ID.
export SLURM_JOBID=$SLURM_JOB_ID                                                                                    # Deprecated. Same as $SLURM_JOB_ID
export SLURM_ARRAY_TASK_ID=${container_indexes[${JOB_CONTAINER_INDEX}]}                        # Task ID.

unmask_filename () {
  replaced="$1"

  if [[ "$replaced" == "\\"* ]]; then
      replaced="${replaced//\\/}"
      echo "${replaced}"
      return 0
  fi

  replaced=$(echo "$replaced" | sed -E "s/(%)(%A)/\1\n\2/g;:a s/(^|[^\n])%A/\1$SLURM_ARRAY_JOB_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%a)/\1\n\2/g;:a s/(^|[^\n])%a/\1$SLURM_ARRAY_TASK_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%j)/\1\n\2/g;:a s/(^|[^\n])%j/\1$SLURM_JOB_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%N)/\1\n\2/g;:a s/(^|[^\n])%N/\1$HOSTNAME/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%n)/\1\n\2/g;:a s/(^|[^\n])%n/\1$JOB_COMPLETION_INDEX/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%t)/\1\n\2/g;:a s/(^|[^\n])%t/\1$SLURM_ARRAY_TASK_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%u)/\1\n\2/g;:a s/(^|[^\n])%u/\1$USER_ID/;ta;s/\n//g")
  replaced=$(echo "$replaced" | sed -E "s/(%)(%x)/\1\n\2/g;:a s/(^|[^\n])%x/\1$SBATCH_JOB_NAME/;ta;s/\n//g")

  replaced="${replaced//%%/%}"

  echo "$replaced"
}

input_file=$(unmask_filename "$SBATCH_INPUT")
output_file=$(unmask_filename "$SBATCH_OUTPUT")
error_path=$(unmask_filename "$SBATCH_ERROR")

/slurm/script


script:
----
#!/bin/bash

#SBATCH --array=1-3%2

echo "now processing task id:: ${SLURM_ARRAY_TASK_ID}"
python /home/slurm/samples/sample_code.py

exit 0



BinaryData
====
```

mwielgus commented:
/lgtm

mwielgus commented:
/unhold
